File: 백준/Gold/15961. 회전 초밥/회전 초밥.py
import sys

n, d, k, c = map(int, input().split())
sushi = [int(sys.stdin.readline()) for _ in range(n)]
s = set()
cnt = [0] * (d + 1)
for i in range(k):
    cnt[sushi[i]] += 1
    if sushi[i] not in s:
        s.add(sushi[i])
start = 0
end = k
ans = len(s)
for i in range(n):
    if end >= n:
        end = 0
    cnt[sushi[start]] -= 1
    cnt[sushi[end]] += 1
    if cnt[sushi[start]] <= 0:
        s.remove(sushi[start])
    if sushi[end] not in s:
        s.add(sushi[end])
    if c in s:
        ans = max(ans, len(s))
    else:
        ans = max(ans, len(s) + 1)

    start += 1
    end += 1

print(ans)

File: 백준/Gold/15961. 회전 초밥/회전 초밥.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int d = scanner.nextInt();
        int k = scanner.nextInt();
        int c = scanner.nextInt();
        int[] sushi = new int[n];
        for (int i = 0; i < n; i++) {
            sushi[i] = scanner.nextInt();
        }
        
        Set<Integer> s = new HashSet<>();
        int[] cnt = new int[d + 1];
        for (int i = 0; i < k; i++) {
            cnt[sushi[i]]++;
            if (!s.contains(sushi[i])) {
                s.add(sushi[i]);
            }
        }
        
        int start = 0;
        int end = k;
        int ans = s.size();
        for (int i = 0; i < n; i++) {
            if (end >= n) {
                end = 0;
            }
            cnt[sushi[start]]--;
            cnt[sushi[end]]++;
            if (cnt[sushi[start]] <= 0) {
                s.remove(sushi[start]);
            }
            if (!s.contains(sushi[end])) {
                s.add(sushi[end]);
            }
            if (s.contains(c)) {
                ans = Math.max(ans, s.size());
            } else {
                ans = Math.max(ans, s.size() + 1);
            }
            
            start++;
            end++;
        }
        
        System.out.println(ans);
    }
}